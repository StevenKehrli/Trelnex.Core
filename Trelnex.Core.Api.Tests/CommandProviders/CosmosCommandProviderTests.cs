using Azure.Identity;
using Microsoft.Azure.Cosmos;
using Microsoft.Azure.Cosmos.Linq;
using Microsoft.Extensions.Configuration;
using Microsoft.Extensions.DependencyInjection;
using Trelnex.Core.Api.CommandProviders;
using Trelnex.Core.Api.Identity;
using Trelnex.Core.Api.Serilog;
using Trelnex.Core.Data;
using Trelnex.Core.Data.Tests.CommandProviders;

namespace Trelnex.Core.Api.Tests.CommandProviders;

[Ignore("Requires a CosmosDB instance.")]
public class CosmosCommandProviderTests : CommandProviderTests
{
    private Container _container = null!;

    [OneTimeSetUp]
    public void TestFixtureSetup()
    {
        // This method is called once prior to executing any of the tests in the fixture.

        // create the service collection
        var services = new ServiceCollection();

        // create the test configuration
        var configuration = new ConfigurationBuilder()
            .AddJsonFile("appsettings.json", optional: true, reloadOnChange: true)
            .AddJsonFile("appsettings.User.json", optional: true, reloadOnChange: true)
            .Build();

        // create a cosmos client for cleanup
        var tokenCredential = new DefaultAzureCredential();

        var endpointUri = configuration
            .GetSection("CosmosCommandProviders:EndpointUri")
            .Value;

        var database = configuration
            .GetSection("CosmosCommandProviders:DatabaseId")
            .Value;

        var container = configuration
            .GetSection("CosmosCommandProviders:Containers:0:ContainerId")
            .Value;

        var cosmosClient = new CosmosClient(
            accountEndpoint: endpointUri,
            tokenCredential: tokenCredential);

        _container = cosmosClient.GetContainer(
            databaseId: database,
            containerId: container);

        var bootstrapLogger = services.AddSerilog(
            configuration,
            "Trelnex.Integration.Tests");

        services
            .AddCredentialFactory(
                configuration,
                bootstrapLogger)
            .AddCosmosCommandProviders(
                configuration,
                bootstrapLogger,
                options => options.Add<ITestItem, TestItem>(
                    typeName: "test-item",
                    validator: TestItem.Validator,
                    commandOperations: CommandOperations.All));

        var serviceProvider = services.BuildServiceProvider();

        // get the command provider
        _commandProvider = serviceProvider.GetRequiredService<ICommandProvider<ITestItem>>();
    }

    [TearDown]
    public async Task TestCleanup()
    {
        // This method is called after each test case is run.

        var feedIterator = _container
            .GetItemLinqQueryable<CosmosItem>()
            .ToFeedIterator();

        while (feedIterator.HasMoreResults)
        {
            var feedResponse = await feedIterator.ReadNextAsync();

            foreach (var item in feedResponse)
            {
                await _container.DeleteItemAsync<CosmosItem>(
                    id: item.id,
                    partitionKey: new PartitionKey(item.partitionKey));
            }
        }
    }

    private record CosmosItem(
        string id,
        string partitionKey);
}
