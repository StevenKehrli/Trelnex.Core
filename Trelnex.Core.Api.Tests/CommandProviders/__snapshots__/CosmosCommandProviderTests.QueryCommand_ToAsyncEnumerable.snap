[
  {
    "Item": {
      "Id": "4ed21fda-4fae-4709-90ca-5c96d85dac78",
      "PartitionKey": "a9b616d7-05e8-4b89-a206-dabf4f19d46a",
      "TypeName": "test-item",
      "CreatedDate": "2025-01-03T07:02:29.910957Z",
      "UpdatedDate": "2025-01-03T07:02:29.910957Z",
      "DeletedDate": null,
      "IsDeleted": null,
      "ETag": "\"8601d6ff-0000-0300-0000-67778b850000\"",
      "Message": "Message #1"
    }
  },
  {
    "Item": {
      "Id": "2467d367-1a5d-4ac0-b760-01850b9ff4f4",
      "PartitionKey": "13e5b99b-893c-4202-a27e-b4aef69f6174",
      "TypeName": "test-item",
      "CreatedDate": "2025-01-03T07:02:29.965735Z",
      "UpdatedDate": "2025-01-03T07:02:29.965735Z",
      "DeletedDate": null,
      "IsDeleted": null,
      "ETag": "\"8601d9ff-0000-0300-0000-67778b860000\"",
      "Message": "Message #2"
    }
  }
]
